core/gtk: add apprt action to show native GUI warning when child exits

jcollie opened the pull request:
Addresses #7649 for the core and GTK. macOS support will need to be added later.

This adds an apprt action to show a native GUI warning of some kind when the child process of a terminal exits.

Also adds a basic GTK implementation of this. In GTK it overlays an Adwaita banner at the bottom of the window (similar to the banner that shows up in at the top of windows in debug builds).
![Screenshot From 2025-07-08 14-49-59](https://github.com/user-attachments/assets/5de4c065-6b35-4ef5-ba26-171c40c0c5ee)
![Screenshot From 2025-07-08 14-50-14](https://github.com/user-attachments/assets/d6259aa1-6eb2-4c7f-9a9b-13dbad706a3e)

jcollie committed:
core/gtk: add apprt action to show native GUI warning when child exits

Addresses #7649 for the core and GTK. macOS support will need to be
added later.

This adds an apprt action to show a native GUI warning of some kind when
the child process of a terminal exits.

Also adds a basic GTK implementation of this. In GTK it overlays an
Adwaita banner at the bottom of the window (similar to the banner that
shows up in at the top of windows in debug builds).

mitchellh reviewed (changes requested):
One code change.

I think stylistically I have to think about this. 

One UX question I have is what does this do to the terminal window if there is text that takes up the full screen? Does the overlay cover it so it's unreadable or does it shrink the terminal area or how is that handled? I'm not able to test this right now.

jcollie commented:
> One code change.
> 
> I think stylistically I have to think about this.
> 
> One UX question I have is what does this do to the terminal window if there is text that takes up the full screen? Does the overlay cover it so it's unreadable or does it shrink the terminal area or how is that handled? I'm not able to test this right now.

Right now the banner will cover up anything at the bottom of the screen. I'm looking at making the banner transparent. Adding a button to dismiss the banner would be easy enough too.

mitchellh commented:
I'd be curious what Gnome Terminal does. I think transparency won't solve things enough because it may be useful to copy text back there. And I'm not sure about dismissing either because it's unclear what state the window is in at that state. This requires some more thought (and is the main reason I haven't pursued it yet). You may want to open a discussion so we can do design discussion in something other than a PR, and hold this PR until then.

jcollie committed:
show child exited: fix macos build

mitchellh commented:
Oh okay, the overlay doesn't impact selection, that's maybe workable... I'll have to try it!

jcollie committed:
show child exited: make GTK banner transparent

jcollie committed:
show child exited: return a boolean if native GUI is shown

If a native GUI is shown by the runtime when a child exits, use the
returned boolean to determine if text should be show in the terminal to
avoid duplicating information.

jcollie commented on src/Surface.zig:
```diff
@@ -1018,6 +1018,14 @@ fn childExited(self: *Surface, info: apprt.surface.Message.ChildExited) void {
             return;
         };
 
+        _ = self.rt_app.performAction(
```
OK, this is implemented. Will update the screenshots in the first message.

jcollie committed:
gtk show_child_edit: use different text for normal/abnormal exit

Without this, the only indication would be the difference between a red
or a green background which would be problematic for users with limited
vision or color blindness.

jcollie commented:
One note, not sure how this affects restoring from a saved state. GTK doesn't do that so I have no way to test. There may need to be some more work that needs to be done to make sure that the appropriate banner is shown again when a surface is restored.

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2502,3 +2502,35 @@ fn gtkStreamError(media_file: *gtk.MediaFile, _: *gobject.ParamSpec, _: ?*anyopa
     media_file.unref();
 }
+
+pub fn showChildExited(self: *Surface, info: apprt.surface.Message.ChildExited) (error{})!bool {
```
Honestly `(error{})!bool` is really funny looking. Couldn't it just be `error{}!bool`?

jcollie replied:
Not sure, but I've had problems trying to put an explicit error set into a function signature before.

jcollie replied:
Yeah that works just fine, not sure why I ran into problems before.

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2502,3 +2502,35 @@ fn gtkStreamError(media_file: *gtk.MediaFile, _: *gobject.ParamSpec, _: ?*anyopa
+    warning_box.as(gtk.Widget).setValign(.end);
+
+    const warning_text = if (info.exit_code == 0)
+        i18n._("⚠️ Process exited normally. Press any key to close the terminal.")
```
It's a bit strange that the normal path still uses the `⚠️` emoji... Maybe ✔️ could work better?

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2502,3 +2502,35 @@ fn gtkStreamError(media_file: *gtk.MediaFile, _: *gobject.ParamSpec, _: ?*anyopa
+
+    const banner_widget = banner.as(gtk.Widget);
+
+    banner_widget.addCssClass("child_exited");
```
Is this CSS class ever used anywhere?

jcollie replied:
Not at the moment.

jcollie replied:
Removed.

pluiedev commented on src/apprt/gtk/Surface.zig:
```diff
@@ -2502,3 +2502,35 @@ fn gtkStreamError(media_file: *gtk.MediaFile, _: *gobject.ParamSpec, _: ?*anyopa
+
+    banner_widget.addCssClass("child_exited");
+
+    if (info.exit_code == 0)
+        banner_widget.addCssClass("child_exited_normally")
+    else
+        banner_widget.addCssClass("child_exited_abnormally");
```
Honestly I probably would just do the dumb route and have two copies of the banner as blueprints, and just hardcode the style and the label text

jcollie commented:
> I don't love the transparency. It really hurts readability. I love the coloring to signify success, but green/red is not accessible for certain variations of color blindness. Does libadwaita abstract this issue away? I'm actually not aware of color blindness being something that is stored in the settings portal.

It's not reflected in the screenshots but one of the last commits changed the verbiage to include whether the exit was normal/abnormal, specifically because of color blindness and screen reader concerns.

This is custom CSS adding the color and transparency.

> I don't understand why the banner is overlaid.

1) It's easy because we already have an overlay.
2) I didn't like the idea of adding a widget to the bottom of the terminal as that would change the height of the window/tab/split or the terminal. That would be kinda janky I think as it might cause a reflow, esp with multiple splits. Tiled window manager behavior would be odd I think.

jcollie committed:
fix error set for function return signature

jcollie committed:
use checkmark for icon if child exits normally

jcollie committed:
remove unused css class from gtk child exited widget

jcollie commented:
FYI, this is what it looks like when running multiple splits:
![image](https://github.com/user-attachments/assets/2dde2f72-e48d-4f12-8485-8443c90162d7)

jcollie committed:
show_child_exited: remove emojis

jcollie commented:
Here's what the close button looks like:

![image](https://github.com/user-attachments/assets/88dfe021-914f-41d2-bf71-d59cf4cc91a5)
![image](https://github.com/user-attachments/assets/0b8a6f7c-b6e8-444b-a3a5-dade3f98dbdb)

jcollie committed:
gtk: GtkBox not needed for show_child_exited

jcollie committed:
gtk: add some comments for show_child_exited

jcollie committed:
gtk: use close button in show_child_exited banner

jcollie committed:
gtk: change child exited verbiage to be clearer and code cleanup

mitchellh commented:
I'm good with the look of this.

mitchellh reviewed (approved):
Approve pending @pluiedev

jcollie committed:
i18n: update translations
